tests/run-make/incr-add-rust-src-component/rmake.rs

```diff
@@ -0,0 +1,39 @@
+// rust-lang/rust#70924: Test that if we add rust-src component in between
+// two incremental compiles, the compiler does not ICE on the second.
+// Remove the rust-src part of the sysroot for the *first* build.
+// Then put in a copy of the rust-src
+// component for the second build, in order to expose the ICE from issue #70924.
+// See https://github.com/rust-lang/rust/pull/72952
+
+//@ needs-symlink
+
+//FIXME(Oneirical): try on test-various
+
+use run_make_support::{path, rfs, rustc};
+
+fn main() {
+    let sysroot = rustc().print("sysroot").run().stdout_utf8();
+    let sysroot = sysroot.trim();
+    rfs::create_dir("fakeroot");
+    symlink_all_entries(&sysroot, "fakeroot");
+    rfs::remove_file("fakeroot/lib");
```

Comment:
Problem [REV1 (2/2)]: unfortunately, there's a whole rabbit hole for symlinks, with different tunnel branches for each platform. The following description is only true for your bog-standard symbolic link ("soft links" that is unofficially called to distinguish against hard links), not junction points, nor hard links.

### On Windows

But young padawan (myself included), the API contract for [`CreateSymlinkW`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinkw) on Windows says:

> To remove a symbolic link, delete the file (using [`DeleteFile`](https://learn.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-deletefilea) or similar APIs) or remove the directory (using [`RemoveDirectory`](https://learn.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-removedirectorya) or similar APIs) depending on what type of symbolic link is used.

Those are the underlying Win32 APIs called by [`std::fs::remove_file`][remove_file] and [`std::fs::remove_dir`][remove_dir] respectively.

If we don't know the kind of symlink on Windows beforehand, we can check it with [`std::os::windows::fs::FileTypeExt`](https://doc.rust-lang.org/std/os/windows/fs/trait.FileTypeExt.html) with [`is_symlink_dir`](https://doc.rust-lang.org/std/os/windows/fs/trait.FileTypeExt.html#tymethod.is_symlink_dir) and [`is_symlink_file`](https://doc.rust-lang.org/std/os/windows/fs/trait.FileTypeExt.html#tymethod.is_symlink_dir). Thank you so much @ChrisDenton!

This `remove_file` and several below it is wrong (on Windows) because it's trying to call `DeleteFileW` on a directory.

### On Linux / macOS

But on Linux/macOS, we need to use [`std::fs::remove_file`][remove_file] because that corresponds to `unlink` on Unix, and there's no distinction for `unlink` with regard to symbolic links in terms of if the symbolic links point to files or directories.

[remove_file]: https://doc.rust-lang.org/std/fs/fn.remove_file.html
[remove_dir]: https://doc.rust-lang.org/std/fs/fn.remove_dir.html

---

Ah but wait, there's another caveat. If you want to interrogate the metadata of the symlink *itself*, you must use [`std::fs::symlink_metadata`](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html) and not [`std::fs::metadata`](https://doc.rust-lang.org/std/fs/fn.metadata.html) which will follow through symlinks to the linked directory/file.

> [`std::fs::FileType`](https://doc.rust-lang.org/std/fs/struct.FileType.html#method.is_symlink): The underlying [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html) struct needs to be retrieved with the [fs::symlink_metadata](https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html) function and not the [fs::metadata](https://doc.rust-lang.org/std/fs/fn.metadata.html) function. The [fs::metadata](https://doc.rust-lang.org/std/fs/fn.metadata.html) function follows symbolic links, so [is_symlink](https://doc.rust-lang.org/std/fs/struct.FileType.html#method.is_symlink) would always return false for the target file.

Reply:
Suggestion: we probably want to implement our very own `remove_symlink` helper that handles these specialness.